### PR TITLE
Feature/offensive positioning

### PR DESCRIPTION
**Pull request by RobZelluf**
- [ ] The tests are passing

### Summary
Give a short summary of what the changes are and why you made them. If applicable, apply screenshots as well. 

### Related issues
If there are related issues, reference them here.

fixes ...  
fixes ...

### Comments for the reviewer
If there are comments for the reviewer, leave them here

### Checklist for the reviewer
This acts as a guide to help you and/or the reviewer to analyze the code. 

**functionality**  
- Is the code is properly tested?
- Is the code working as expected
- Is the code is understandable and easy to read
- Are proper abstractions made? (no duplicate code, nice use of functions/objects)

**style**  
- Variables/objects/classes have proper names
- Proper use of variables: nu unused vars, no magic numbers
- No possible memory leaks. Mutexes properly used when multithreading? possible pointer issues?
- Are smart pointers used?
- unused variables?
- magic numbers?
- well documented?
- No unused imports?
- No leftover code / unused comments
- Proper indentation
- Proper use of namespaces

### CODE REVIEW

File: roboteam_ai/src/control/ControlUtils.cpp [hunk 1]
```diff
@@ -93,18 +105,11 @@ bool ControlUtils::hasClearVision(int fromID, int towardsID, roboteam_msgs::Worl
...
-    Vector2 c = n*(n.dot(pa)/n.dot(n));
-    Vector2 d = pa - c;
-    Vector2 A = (pointToCheck - lineStart).project(lineStart, lineEnd);
-    Vector2 B = (pointToCheck - lineEnd).project(lineEnd, lineStart);
-    if ((A.length() + B.length()) > n.length()) {
-        return fmin(pa.length(), (lineEnd - pointToCheck).length());
-    }
-    else return d.length();
+    Vector2 projectionPoint = pointToCheck.project(lineStart,lineEnd);
+    return (projectionPoint - pointToCheck).length();
```
Comment (RobZelluf) [inline roboteam_ai/src/control/ControlUtils.cpp]: @rolfvdhulst could you quickly look at this? I changed it so that it works for me, but I think you are actually using this code (and know what it does)?

File: roboteam_ai/src/utilities/Coach.cpp [hunk 2]
```diff
@@ -116,7 +116,7 @@ bool Coach::isRobotBehindBallToPosition(double distanceBehindBall, const Vector2
     Vector2 behindBallPosition = getPositionBehindBallToPosition(distanceBehindBall, position);
     Vector2 deltaBall = behindBallPosition - ball;
 
-    double angleMargin = 0.12;
+    double angleMargin = 0.02;
```
Comment (RobZelluf) [inline roboteam_ai/src/utilities/Coach.cpp]: This should be moved to constants

File: roboteam_ai/src/control/ControlUtils.cpp (hunk 1 above)
Comment (rolfvdhulst) [reply roboteam_ai/src/control/ControlUtils.cpp]: Well as is, this is wrong, as this will extend your line indefinitely  through the points lineStart and linEnd, and don't work with the line segment (so this is just distance to line). The previous version was also broken. There is a working version in the defensiveCoach branch.

I suggest you put a ToDo /warning here for now and that we merge it through another PR (I think @mrlukasbos has the same working one on another branch too?)

File: roboteam_ai/src/conditions/ShotAtGoal.h [hunk 3]
```diff
@@ -0,0 +1,27 @@
+//
+// Created by robzelluf on 3/14/19.
+//
+
+#ifndef ROBOTEAM_AI_SHOTATGOAL_H
```
Comment (rolfvdhulst) [inline roboteam_ai/src/conditions/ShotAtGoal.h]: Could you maybe refactor the class to be named hasClearShot or something that is a bit less vague? ShotAtGoal also could mean that you already have shot the ball at goal

File: roboteam_ai/src/utilities/OffensiveCoach.cpp [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+//
+
+#include <roboteam_ai/src/interface/widget.h>
+#include <roboteam_ai/src/interface/drawer.h>
+#include "OffensiveCoach.h"
+
+namespace rtt {
+namespace ai {
+namespace coach {
+
```
Comment (rolfvdhulst) [inline roboteam_ai/src/utilities/OffensiveCoach.cpp]: the 2 max variables should be constants, IMO

File: roboteam_ai/src/utilities/OffensiveCoach.cpp [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+double OffensiveCoach::calculateCloseToGoalScore(Vector2 position) {
+    double distanceFromGoal = (Field::get_their_goal_center() - position).length();
+
+    double score = exp(-0.1 * distanceFromGoal);
+    return score;
+}
+
+double OffensiveCoach::calculateShotAtGoalScore(Vector2 position, roboteam_msgs::World world) {
+    double safelyness = 3;
```
Comment (rolfvdhulst) [inline roboteam_ai/src/utilities/OffensiveCoach.cpp]: Can we use another word like safeDistanceFactor or so? This word just feels wrong.

File: roboteam_ai/src/utilities/OffensiveCoach.cpp [hunk 6]
```diff
@@ -0,0 +1,317 @@
+//
```
Comment (rolfvdhulst) [inline roboteam_ai/src/utilities/OffensiveCoach.cpp]: I like the use of exponential functions to model probabilities

File: roboteam_ai/src/ApplicationManager.cpp [hunk 7]
```diff
@@ -82,6 +82,8 @@ void ApplicationManager::runOneLoopCycle() {
         strategy = factory.getTree(BTFactory::getCurrentTree());
         Status status = strategy->tick();
         this->notifyTreeStatus(status);
+
+        rtt::ai::coach::OffensiveCoach::calculateNewPositions();
```
Comment (mrlukasbos) [inline roboteam_ai/src/ApplicationManager.cpp]: so you are always calculating positions, regardless of the strategy. This feels weird. What positions are you calculating?

File: roboteam_ai/src/conditions/IsBeingPassedTo.cpp [hunk 8]
```diff
@@ -13,7 +13,8 @@ IsBeingPassedTo::IsBeingPassedTo(std::string name, bt::Blackboard::Ptr blackboar
 void IsBeingPassedTo::onInitialize() {};
 
 IsBeingPassedTo::Status IsBeingPassedTo::onUpdate() {
-    if (coach::Coach::getRobotBeingPassedTo() == static_cast<int>(getRobotFromProperties(properties)->id)) {
+    robot = getRobotFromProperties(properties);
```
Comment (mrlukasbos) [inline roboteam_ai/src/conditions/IsBeingPassedTo.cpp]: this line is not necessary

File: roboteam_ai/src/conditions/ShotAtGoal.cpp [hunk 9]
```diff
@@ -0,0 +1,32 @@
+//
+// Created by robzelluf on 3/14/19.
+//
+
+#include "ShotAtGoal.h"
+
+namespace rtt{
+namespace ai {
+
+ShotAtGoal::ShotAtGoal(std::string name, bt::Blackboard::Ptr blackboard)
```
Comment (mrlukasbos) [inline roboteam_ai/src/conditions/ShotAtGoal.cpp]: The name of this condition is a little bit weird. a shot at which goal? Does it return succes when a shot at goal happens or when it is possible for an attacker (and an attacker from which team then?)

File: roboteam_ai/src/conditions/ShotAtGoal.cpp [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+        :Condition(std::move(name), std::move(blackboard)) {};
+
+void ShotAtGoal::initialize() {};
+
+ShotAtGoal::Status ShotAtGoal::update() {
+    robot = getRobotFromProperties(properties);
```
Comment (mrlukasbos) [inline roboteam_ai/src/conditions/ShotAtGoal.cpp]: use onUpdate() and remove this line

File: roboteam_ai/src/conditions/ShotAtGoal.h (hunk 3 above)
Comment (mrlukasbos) [reply roboteam_ai/src/conditions/ShotAtGoal.h]: agreed

File: roboteam_ai/src/interface/drawer.cpp [hunk 11]
```diff
@@ -1,3 +1,5 @@
+#include <utility>
```
Comment (mrlukasbos) [inline roboteam_ai/src/interface/drawer.cpp]: not needed

File: roboteam_ai/src/jsons/roles/MiddleFielder.json [hunk 12]
```diff
@@ -0,0 +1 @@
+{"data": {"trees": [{"nodes": {"ztoudkgn68rp93s6x": {"properties": {"ball": "true", "margin": "0.12"}, "id": "ztoudkgn68rp93s6x", "title": "IsCloseToPoint"}, "2kqqtoiuvkcgu": {"id": "2kqqtoiuvkcgu", "title": "SideAttacker"}, "21iy52sewbnusbx": {"children": ["ztoudkgn68rp93s6x", "3t4kcbwinbjmigw1"], "id": "21iy52sewbnusbx", "title": "Selector"}, "ea54a5hzyqx2o3bt": {"id": "ea54a5hzyqx2o3bt", "title": "Pass"}, "3t4kcbwinbjmigw1": {"properties": {"goToBall": "true", "canGoOutsideField": "false", "avoidBall": "false"}, "id": "3t4kcbwinbjmigw1", "title": "BasicGoToPos"}, "x8xkl0lh5k63kr7u": {"children": ["21iy52sewbnusbx", "ea54a5hzyqx2o3bt"], "id": "x8xkl0lh5k63kr7u", "title": "MemSequence"}, "xuqhz22hql231ry": {"children": ["x8xkl0lh5k63kr7u", "2kqqtoiuvkcgu"], "id": "xuqhz22hql231ry", "title": "MemSequence"}}, "root": "xuqhz22hql231ry", "title": "MiddleFielder"}]}, "name": "MiddleFielder"}
```
Comment (mrlukasbos) [inline roboteam_ai/src/jsons/roles/MiddleFielder.json]: refactor the filename to midfielder

File: roboteam_ai/src/skills/Pass.cpp [hunk 13]
```diff
@@ -23,14 +23,20 @@ Pass::Status Pass::onUpdate() {
 
     switch(currentProgress) {
         case Progression::POSITIONING: {
-            if (!coach::Coach::isRobotBehindBallToPosition(0.30, robotToPassTo->pos, robot->pos)) {
-                goToType = GoToType::NUMERIC_TREES;
-                targetPos = Coach::getPositionBehindBallToPosition(0.30, robotToPassTo->pos);
-            } else if (!World::ourBotHasBall(robot->id)) {
+            if (!coach::Coach::isRobotBehindBallToPosition(0.40, robotToPassTo->pos, robot->pos)) {
+                std::cout << "Going behind ball" << std::endl;
```
Comment (mrlukasbos) [inline roboteam_ai/src/skills/Pass.cpp]: There are print statements which should be removed. Also... why did you change passing in this PR?

File: roboteam_ai/src/utilities/Coach.cpp [hunk 14]
```diff
@@ -265,8 +263,8 @@ void Coach::resetPass() {
 int Coach::initiatePass() {
     resetPass();
 
-    // TODO: More logic to decide which robot to pass to. Possibly split initiate in initiate and findRobotToPassTo
-    int robotBeingPassedTo = Field::getRobotClosestToGoal(true, false);
+    int robotBeingPassedTo = coach::OffensiveCoach::getBestStrikerID();
+    std::cout << "Pass to robot " << robotBeingPassedTo << std::endl;
```
Comment (mrlukasbos) [inline roboteam_ai/src/utilities/Coach.cpp]: remove this print statement

File: roboteam_ai/src/utilities/OffensiveCoach.cpp [hunk 15]
```diff
@@ -0,0 +1,312 @@
...
+#include <roboteam_ai/src/interface/widget.h>
+#include <roboteam_ai/src/interface/drawer.h>
+#include "OffensiveCoach.h"
+
+namespace rtt {
+namespace ai {
+namespace coach {
+
+int OffensiveCoach::maxPositions = 4;
+double OffensiveCoach::maxDistanceFromBall = 6.0;
```
Comment (mrlukasbos) [inline roboteam_ai/src/utilities/OffensiveCoach.cpp]: magic numbers

File: roboteam_ai/src/utilities/OffensiveCoach.cpp (hunk 5 above)
Comment (mrlukasbos) [reply roboteam_ai/src/utilities/OffensiveCoach.cpp]: agreed

File: roboteam_ai/src/utilities/OffensiveCoach.cpp [hunk 16]
```diff
@@ -0,0 +1,312 @@
...
+    return score;
+}
+
+double OffensiveCoach::calculateShotAtGoalScore(Vector2 position, roboteam_msgs::World world) {
+    double safelyness = 3;
+    while (safelyness > 0) {
+            if (control::ControlUtils::clearLine(position, Field::get_their_goal_center(), world, safelyness)) {
+                break;
+            }
+            safelyness -= 0.5;
```
Comment (mrlukasbos) [inline roboteam_ai/src/utilities/OffensiveCoach.cpp]: magic

File: roboteam_ai/src/utilities/OffensiveCoach.cpp [hunk 17]
```diff
@@ -0,0 +1,312 @@
...
+        safelyness -= 0.5;
+    }
+
+    return 1 - exp(-safelyness);
+}
+
+double OffensiveCoach::calculateDistanceToOpponentsScore(Vector2 position, roboteam_msgs::World world) {
+    shared_ptr<roboteam_msgs::WorldRobot> closestRobot = World::getRobotClosestToPoint(world.them, position);
+    if (closestRobot) {
+        double distance = (position - closestRobot->pos).length();
```
Comment (mrlukasbos) [inline roboteam_ai/src/utilities/OffensiveCoach.cpp]: the whole safelyness is magic

File: roboteam_ai/src/utilities/OffensiveCoach.cpp [hunk 18]
```diff
@@ -0,0 +1,312 @@
...
+    double behindBallScore = position.x < world.ball.pos.x ? 0.7 : 1.0;
+    double distanceFromCornerScore = calculateDistanceFromCorner(position, field);
+
+    double score = 2 * closeToGoalScore + 2 * shotAtGoalScore + passLineScore + closestOpponentScore
+            + distanceFromBallScore + behindBallScore + distanceFromCornerScore;
+
+    return score;
+}
+
+void OffensiveCoach::calculateNewPositions() {
```
Comment (mrlukasbos) [inline roboteam_ai/src/utilities/OffensiveCoach.cpp]: this function is way too complex. split it up into multiple smaller ones

File: roboteam_ai/src/utilities/OffensiveCoach.cpp [hunk 19]
```diff
@@ -0,0 +1,312 @@
...
+    roboteam_msgs::GeometryFieldSize field = Field::get_field();
+    double closeToGoalScore = calculateCloseToGoalScore(position);
+    double shotAtGoalScore = calculateShotAtGoalScore(position, world);
+    double passLineScore = calculatePassLineScore(position, world);
+    double closestOpponentScore = calculateDistanceToOpponentsScore(position, world);
+    double distanceFromBallScore = calculateDistanceFromBallScore(position, field, world.ball);
+    double behindBallScore = position.x < world.ball.pos.x ? 0.7 : 1.0;
+    double distanceFromCornerScore = calculateDistanceFromCorner(position, field);
+
+    double score = 2 * closeToGoalScore + 2 * shotAtGoalScore + passLineScore + closestOpponentScore
```
Comment (mrlukasbos) [inline roboteam_ai/src/utilities/OffensiveCoach.cpp]: magic

File: roboteam_ai/src/utilities/OffensiveCoach.cpp [hunk 20]
```diff
@@ -0,0 +1,312 @@
...
+        offensivePositions.erase(offensivePositions.begin() + maxPositions, offensivePositions.end());
+    }
+    drawOffensivePoints();
+}
+
+bool OffensiveCoach::compareByScore(OffensivePosition position1, OffensivePosition position2) {
+    return position1.score > position2.score;
+}
+
+Vector2 OffensiveCoach::calculatePositionForRobot(std::shared_ptr<roboteam_msgs::WorldRobot> robot) {
```
Comment (mrlukasbos) [inline roboteam_ai/src/utilities/OffensiveCoach.cpp]: this method is too complex. please refactor

File: roboteam_ai/src/utilities/OffensiveCoach.cpp [hunk 21]
```diff
@@ -0,0 +1,312 @@
...
+        double currentDistance;
+        OffensiveCoach::OffensivePosition newRobotPosition;
+        for (auto &position : offensivePositions) {
+            currentDistance = (position.position - robot->pos).length();
+            if (currentDistance < distance) {
+                distance = currentDistance;
+                newRobotPosition = position;
+            }
+        }
+        if (distance == 999) {
```
Comment (mrlukasbos) [inline roboteam_ai/src/utilities/OffensiveCoach.cpp]: INT_MAX

File: roboteam_ai/src/ApplicationManager.cpp (hunk 7 above)
Comment (RobZelluf) [reply roboteam_ai/src/ApplicationManager.cpp]: Just generally good offensive positions, good positions to receive the ball and shoot at the goal, and preferably a bit close to the opponents goal

File: roboteam_ai/src/conditions/IsBeingPassedTo.cpp (hunk 8 above)
Comment (RobZelluf) [reply roboteam_ai/src/conditions/IsBeingPassedTo.cpp]: Done

File: roboteam_ai/src/conditions/ShotAtGoal.cpp (hunk 9 above)
Comment (RobZelluf) [reply roboteam_ai/src/conditions/ShotAtGoal.cpp]: Renamed

File: roboteam_ai/src/conditions/ShotAtGoal.cpp (hunk 10 above)
Comment (RobZelluf) [reply roboteam_ai/src/conditions/ShotAtGoal.cpp]: Done

File: roboteam_ai/src/conditions/ShotAtGoal.h (hunk 3 above)
Comment (RobZelluf) [reply roboteam_ai/src/conditions/ShotAtGoal.h]: Done

File: roboteam_ai/src/interface/drawer.cpp (hunk 11 above)
Comment (RobZelluf) [reply roboteam_ai/src/interface/drawer.cpp]: Done

File: roboteam_ai/src/jsons/roles/MiddleFielder.json (hunk 12 above)
Comment (RobZelluf) [reply roboteam_ai/src/jsons/roles/MiddleFielder.json]: Done

File: roboteam_ai/src/skills/Pass.cpp (hunk 13 above)
Comment (RobZelluf) [reply roboteam_ai/src/skills/Pass.cpp]: Done. I wanted to test the positioning with passing, but it wasn't working properly so I had to change some things

File: roboteam_ai/src/utilities/Coach.cpp (hunk 14 above)
Comment (RobZelluf) [reply roboteam_ai/src/utilities/Coach.cpp]: Done

File: roboteam_ai/src/utilities/OffensiveCoach.cpp (hunk 6 above)
Comment (RobZelluf) [reply roboteam_ai/src/utilities/OffensiveCoach.cpp]: Thanks!

File: roboteam_ai/src/utilities/OffensiveCoach.cpp (hunk 4 above)
Comment (RobZelluf) [reply roboteam_ai/src/utilities/OffensiveCoach.cpp]: Just discussed this with @ThijsLuttikhuis today, in our opinion not everything should just go to the constants file. These to variables are very specific to the offensive coach, so I think it should be fine if they're located here.

File: roboteam_ai/src/utilities/OffensiveCoach.cpp (hunk 15 above)
Comment (RobZelluf) [reply roboteam_ai/src/utilities/OffensiveCoach.cpp]: https://cdn.drawception.com/images/panels/2018/5-14/EC5N43PsAP-2.png

Comment (RobZelluf) [reply roboteam_ai/src/utilities/OffensiveCoach.cpp]: But as mentioned before here, since these are very specific to the OffensiveCoach I am not sure whether it makes a lot of sense to put them in constants. They are not used anywhere else and they're still specified in the beginning of the file

File: roboteam_ai/src/utilities/Coach.cpp (hunk 14 above)
Comment (RobZelluf) [reply roboteam_ai/src/utilities/Coach.cpp]: Done

File: roboteam_ai/src/skills/Pass.cpp (hunk 13 above)
Comment (RobZelluf) [reply roboteam_ai/src/skills/Pass.cpp]: Done. I wanted to test the positioning using passing, which wasn't working properly then, so I changed it

File: roboteam_ai/src/utilities/OffensiveCoach.cpp (hunk 6 above)
Comment (RobZelluf) [reply roboteam_ai/src/utilities/OffensiveCoach.cpp]: Thanks!

File: roboteam_ai/src/utilities/OffensiveCoach.cpp (hunk 4 above)
Comment (RobZelluf) [reply roboteam_ai/src/utilities/OffensiveCoach.cpp]: Since they are very specific to the OffensiveCoach, I think it should be fine that they stay here

File: roboteam_ai/src/utilities/OffensiveCoach.cpp (hunk 15 above)
Comment (RobZelluf) [reply roboteam_ai/src/utilities/OffensiveCoach.cpp]: Yeah true, but needed, and too specific to be moved to constants

File: roboteam_ai/src/utilities/OffensiveCoach.cpp (hunk 5 above)
Comment (RobZelluf) [reply roboteam_ai/src/utilities/OffensiveCoach.cpp]: Done

File: roboteam_ai/src/utilities/OffensiveCoach.cpp (hunk 17 above)
Comment (RobZelluf) [reply roboteam_ai/src/utilities/OffensiveCoach.cpp]: I agree, but I also think it's an okay method. It's just a measure for the degree the line is not obstructed by other objects

File: roboteam_ai/src/utilities/OffensiveCoach.cpp (hunk 19 above)
Comment (RobZelluf) [reply roboteam_ai/src/utilities/OffensiveCoach.cpp]: Well yeah in that sense the whole scoring is magic

File: roboteam_ai/src/utilities/OffensiveCoach.cpp (hunk 21 above)
Comment (RobZelluf) [reply roboteam_ai/src/utilities/OffensiveCoach.cpp]: Done

File: roboteam_ai/src/utilities/OffensiveCoach.cpp (hunk 20 above)
Comment (RobZelluf) [reply roboteam_ai/src/utilities/OffensiveCoach.cpp]: Done

File: roboteam_ai/src/coach/CoachHeuristics.cpp [hunk 22]
```diff
@@ -0,0 +1,91 @@
...
+
+#include "CoachHeuristics.h"
+
+namespace rtt {
+namespace ai {
+namespace coach {
+
+double CoachHeuristics::maxDistanceFromBall = 6.0;
+
+double CoachHeuristics::calculateCloseToGoalScore(Vector2 position) {
```
Comment (mrlukasbos) [inline roboteam_ai/src/coach/CoachHeuristics.cpp]: tbh I would like a comment above each function to shortly say what it does

File: roboteam_ai/src/coach/CoachHeuristics.cpp [hunk 23]
```diff
@@ -0,0 +1,91 @@
...
+    double distanceFromGoal = (Field::get_their_goal_center() - position).length();
+
+    double score = exp(-0.1 * distanceFromGoal);
+    return score;
+}
+
+double CoachHeuristics::calculateShotAtGoalScore(Vector2 position, roboteam_msgs::World world) {
```
Comment (mrlukasbos) [inline roboteam_ai/src/coach/CoachHeuristics.cpp]: There are quite some magic constants here. I think it could help if you define the scores as a constant somewhere (for example inside the function would be fine) to clearly show what the values mean.

File: roboteam_ai/src/coach/CoachHeuristics.h [hunk 24]
```diff
@@ -0,0 +1,34 @@
...
+
+#include <roboteam_utils/Vector2.h>
+#include <roboteam_ai/src/control/ControlUtils.h>
+#include <roboteam_ai/src/utilities/Field.h>
+
+namespace rtt {
+namespace ai {
+namespace coach {
+
+class CoachHeuristics {
```
Comment (mrlukasbos) [inline roboteam_ai/src/coach/CoachHeuristics.h]: This class does not need to be static

File: roboteam_ai/src/coach/CoachHeuristics.h [hunk 25]
```diff
@@ -0,0 +1,34 @@
...
+private:
+    static double maxDistanceFromBall;
```
Comment (mrlukasbos) [inline roboteam_ai/src/coach/CoachHeuristics.h]: what does this value do? is it not a constant? why not make it `const double`?

File: roboteam_ai/src/coach/CoachHeuristics.h (hunk 24 above)
Comment (mrlukasbos) [reply roboteam_ai/src/coach/CoachHeuristics.h]: But if you keep it like this it's fine for now

File: roboteam_ai/src/coach/GeneralPositionCoach.cpp [hunk 26]
```diff
@@ -39,13 +39,13 @@ Vector2 GeneralPositionCoach::getPositionBehindBallToPosition(double distanceBeh
...
-bool GeneralPositionCoach::isRobotBehindBallToGoal(double distanceBehindBall, bool ourGoal, const Vector2 &robotPosition) {
+bool GeneralPositionCoach::isRobotBehindBallToGoal(double distanceBehindBall, bool ourGoal, const Vector2 &robotPosition, double angleMargin) {
     const Vector2 &goal = (ourGoal ? Field::get_our_goal_center : Field::get_their_goal_center)();
-    return isRobotBehindBallToPosition(distanceBehindBall, goal, robotPosition);
+    return isRobotBehindBallToPosition(distanceBehindBall, goal, robotPosition, angleMargin);
 }
 
 bool GeneralPositionCoach::isRobotBehindBallToRobot(double distanceBehindBall, bool ourRobot, const unsigned int &robotID,
-        const Vector2 &robotPosition) {
+        const Vector2 &robotPosition, double angleMargin) {
```
Comment (mrlukasbos) [inline roboteam_ai/src/coach/GeneralPositionCoach.cpp]: `angleMargin` is not used here

File: roboteam_ai/src/coach/GeneralPositionCoach.h [hunk 27]
```diff
@@ -7,6 +7,7 @@
 #define ROBOTEAM_AI_GENERALPOSITIONCOACH_H
 
 #include "roboteam_utils/Vector2.h"
+#include <roboteam_ai/src/utilities/Constants.h>
```
Comment (mrlukasbos) [inline roboteam_ai/src/coach/GeneralPositionCoach.h]: These constants could be defined in the H file itself now

File: roboteam_ai/src/coach/OffensiveCoach.cpp [hunk 28]
```diff
@@ -0,0 +1,250 @@
...
+            if ((it->position - robotPosition.second.position).length() < Constants::ATTACKER_DISTANCE()) {
+                offensivePositions.erase(it);
+                break;
+            }
+        }
+        it++;
+    }
+}
+
+OffensiveCoach::OffensivePosition OffensiveCoach::calculateRandomPosition(double xStart, double xEnd, double yStart,
```
Comment (mrlukasbos) [inline roboteam_ai/src/coach/OffensiveCoach.cpp]: add some comments

File: roboteam_ai/src/coach/OffensiveCoach.cpp [hunk 29]
```diff
@@ -0,0 +1,250 @@
...
+        offensivePositions.erase(offensivePositions.begin() + maxPositions, offensivePositions.end());
+    }
+    drawOffensivePoints();
+}
+
+void OffensiveCoach::compareToCurrentPositions(const OffensiveCoach::OffensivePosition &position) {
+    OffensivePosition bestPos = position;
+    auto it = offensivePositions.begin();
+    while (it < offensivePositions.end()) {
+        if ((position.position - it->position).length() < Constants::OFFENSIVE_POSITION_DISTANCE()) {//Constants::ATTACKER_DISTANCE()) {
```
Comment (mrlukasbos) [inline roboteam_ai/src/coach/OffensiveCoach.cpp]: remove this comment

File: roboteam_ai/src/coach/OffensiveCoach.cpp [hunk 30]
```diff
@@ -0,0 +1,250 @@
...
+        calculateNewRobotPositions(robot);
+        if ((robotPositions.find(robot->id) == robotPositions.end())) { //not in there
+            return calculatePositionForRobot(robot);
+        } else {
+            return robotPositions[robot->id].position;
+        }
+    }
+}
+
+Vector2 OffensiveCoach::getClosestOffensivePosition(const shared_ptr<roboteam_msgs::WorldRobot> &robot) {
```
Comment (mrlukasbos) [inline roboteam_ai/src/coach/OffensiveCoach.cpp]: quite sure you can use hungarian which kinda does the same. Also, you could just simplify this a lot by comparing `robot->pos.dist(position.position)` for each robot.

File: roboteam_ai/src/coach/OffensiveCoach.cpp [hunk 31]
```diff
@@ -0,0 +1,250 @@
...
+void OffensiveCoach::calculateNewRobotPositions(std::shared_ptr<roboteam_msgs::WorldRobot> robot) {
+    OffensiveCoach::OffensivePosition currentRobotPosition = robotPositions[robot->id];
+
+    Vector2 currentPosition = currentRobotPosition.position;
+    double currentScore = CoachHeuristics::calculatePositionScore(currentPosition);
+
+    OffensivePosition newPosition;
+    bool foundNewPosition = false;
+
+    for (int xDiff : {-2, -1, 0, 1, 2}) {
```
Comment (mrlukasbos) [inline roboteam_ai/src/coach/OffensiveCoach.cpp]: this looks pretty ugly

File: roboteam_ai/src/coach/OffensiveCoach.cpp [hunk 32]
```diff
@@ -0,0 +1,250 @@
...
+        for (int yDiff : {-2, -1, 0, 1, 2}) {
+            newPosition.position.x = currentPosition.x + 0.01 * xDiff;
+            newPosition.position.y = currentPosition.y + 0.01 * yDiff;
```
Comment (mrlukasbos) [inline roboteam_ai/src/coach/OffensiveCoach.cpp]: magic .01

File: roboteam_ai/src/coach/OffensiveCoach.cpp (hunk 31 above)
Comment (RobZelluf) [reply roboteam_ai/src/coach/OffensiveCoach.cpp]: Yeah I know, but would not know how to make it much nicer

Comment (RobZelluf) [reply roboteam_ai/src/coach/OffensiveCoach.cpp]: Nvm, made it nicer
